postgis/lwgeom_generate_grid.c

```diff
@@ -0,0 +1,430 @@
+{
+	if (!state || state->done) return false;
+	/* Move up one row */
+	state->j++;
+	/* Off the end, increment column counter, reset row counter back to (appropriate) minimum */
+	if (state->j > state->row_max)
+	{
+		state->i++;
+		state->j = state->row_min;
+	}
+	/* Column counter over max means we have used up all combinations */
+	if (state->i > state->column_max)
+	{
+		state->done = true;
+	}
+	return !state->done;
+}
+
+/**
+* ST_HexagonGrid(size float8, bounds geometry)
+* ST_SquareGrid(size float8, bounds geometry)
+*/
+PG_FUNCTION_INFO_V1(ST_ShapeGrid);
+Datum ST_ShapeGrid(PG_FUNCTION_ARGS)
+{
+	FuncCallContext *funcctx;
+	MemoryContext oldcontext, newcontext;
+
+	GSERIALIZED *gbounds;
+	GeometryGridState *state;
+
+	LWGEOM *lwgeom;
+	bool isnull[3] = {0,0,0}; /* needed to say no value is null */
+	Datum tuple_arr[3]; /* used to construct the composite return value */
+	HeapTuple tuple;
+	Datum result; /* the actual composite return value */
+
+	if (SRF_IS_FIRSTCALL())
+	{
+		const char *func_name;
+		double bounds_width, bounds_height;
+		char gbounds_is_empty;
+		GBOX bounds;
+		double size;
+		funcctx = SRF_FIRSTCALL_INIT();
+
+		/* Get a local copy of the bounds we are going to fill with shapes */
+		gbounds = PG_GETARG_GSERIALIZED_P(1);
+		size = PG_GETARG_FLOAT8(0);
+
+		gbounds_is_empty = (gserialized_get_gbox_p(gbounds, &bounds) == LW_FAILURE);
+		bounds_width = bounds.xmax - bounds.xmin;
+		bounds_height = bounds.ymax - bounds.ymin;
+
+		/* quick opt-out if we get nonsensical inputs  */
+		if (size <= 0.0 || gbounds_is_empty ||
+		    bounds_width <= 0.0 || bounds_height <= 0.0)
+		{
+			funcctx = SRF_PERCALL_SETUP();
+			SRF_RETURN_DONE(funcctx);
+		}
+
+		newcontext = funcctx->multi_call_memory_ctx;
+		oldcontext = MemoryContextSwitchTo(newcontext);
+
+		/*
+		* Support both hexagon and square grids with one function,
+		* by checking the calling signature up front.
+		*/
+		func_name = get_func_name(fcinfo->flinfo->fn_oid);
+		if (strcmp(func_name, "st_hexagongrid") == 0)
+		{
+			state = (GeometryGridState*)hexagon_grid_state(size, &bounds, gserialized_get_srid(gbounds));
+		}
+		else if (strcmp(func_name, "st_squaregrid") == 0)
+		{
+			state = (GeometryGridState*)square_grid_state(size, &bounds, gserialized_get_srid(gbounds));
+		}
+		else
+		{
+			ereport(ERROR, (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+				errmsg("%s called from unsupported functional context '%s'", __func__, func_name)));
+		}
+
+		funcctx->user_fctx = state;
+
+		/* get tuple description for return type */
+		if (get_call_result_type(fcinfo, 0, &funcctx->tuple_desc) != TYPEFUNC_COMPOSITE)
+		{
+			ereport(ERROR, (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+				errmsg("set-valued function called in context that cannot accept a set")));
+		}
+
+		BlessTupleDesc(funcctx->tuple_desc);
+		MemoryContextSwitchTo(oldcontext);
+	}
+
+	/* stuff done on every call of the function */
+	funcctx = SRF_PERCALL_SETUP();
+	newcontext = funcctx->multi_call_memory_ctx;
```

Comment:
I guess I'm missing something but I don't know why this assignment is necessary if `newcontext` isn't used after it.

Reply:
I guess it might be copied from `LWGEOM_dumppoints` which also does a similar dead store.

Reply:
yeah, that's true